File: php/src/Wechat.php
<?php
/**
 * 微信公众号、企业号模板
 *
 * 说明：基于当前版本的微信公众号和企业号
 * @author ChenKang <ck@kchen.cn>
 * @version 1.0
 */

require_once __DIR__ . '/MsgCrypt.php';

class Wechat
{
    /**
     * xml模板字符串
     *
     * @var string
     */
    protected $templateText        = "<xml><ToUserName><![CDATA[%s]]></ToUserName><FromUserName><![CDATA[%s]]></FromUserName><CreateTime>%s</CreateTime><MsgType><![CDATA[text]]></MsgType><Content><![CDATA[%s]]></Content></xml>";
    protected $templateNewsBegin   = "<xml><ToUserName><![CDATA[%s]]></ToUserName><FromUserName><![CDATA[%s]]></FromUserName><CreateTime>%s</CreateTime><MsgType><![CDATA[news]]></MsgType><ArticleCount>%s</ArticleCount><Articles>";
    protected $templateNewsContent = "<item><Title><![CDATA[%s]]></Title><Description><![CDATA[%s]]></Description><PicUrl><![CDATA[%s]]></PicUrl><Url><![CDATA[%s]]></Url></item>";
    protected $templateNewsEnd     = "</Articles></xml>";

    /**
     * 调试模式，将错误通过文本消息回复显示
     *
     * @var boolean
     */
    private $debug;

    /**
     * 以数组的形式保存微信服务器每次发来的请求
     *
     * @var array
     */
    public $request;
    
    /**
     * 加密、解密、验证URl
     *
     * @var MsgCrypt
     */
    private $msgCryptor;

    /**
     * 消息是否加密
     *
     * @var boolean
     */
    private $encrypted = false;

    /**
     * 微信请求时的GET参数，加密消息时要用到
     *
     * @var string
     */
    private $timestamp, $nonce, $msg_signature;

    /**
     * 可接受的消息类型和事件类型数组
     *
     * @var array
     */
    protected $msgTypeArray   = array('text', 'image', 'voice', 'video', 'shortvideo', 'link', 'location', 'event');
    protected $eventTypeArray = array('subscribe','unsubscribe','click','view','location','scan','scancode_push','scancode_waitmsg','pic_sysphoto','pic_photo_or_album','pic_weixin','location_select','enter_agent','batch_job_result');


    /**
     * 初始化，判断此次请求是否为验证请求，并以数组形式保存
     *
     * @param string $token 验证信息
     * @param boolean $debug 调试模式，默认为关闭
     */
    public function __construct($token, $aeskey = '', $appid = '', $debug = FALSE) 
    {

        $this->debug = $debug;

        // 获取请求参数
        $this->timestamp     = $_GET['timestamp'];
        $this->nonce         = $_GET['nonce'];
        if (!$this->debug && (empty($this->timestamp) || empty($this->nonce))) {
            exit("请通过微信访问");
        }

        // 判断消息是否加密
        if (isset($_GET['msg_signature'])) {
            $this->msg_signature = $_GET['msg_signature'];
            $this->encrypted = TURE;
        } else {
            $this->encrypted = FALSE;
        }

        // 判断公众号类型
        if (isset($_GET['signature'])) {
            $wechatType = 1;
            $signature = $_GET['signature'];
        } else {
            $wechatType = 2;
        }

        // 其他参数
        $echostr     = isset($_GET['echostr']) ? $_GET['echostr'] : '';
        $requestData = isset($GLOBALS['HTTP_RAW_POST_DATA']) ? $GLOBALS['HTTP_RAW_POST_DATA'] : file_get_contents("php://input");

        $this->msgCryptor = new MsgCrypt($token, $aeskey, $appid);

        // 验证TOKEN
        if ($this->debug == FALSE) {
            $validResult = TRUE;
            if ($wechatType == 1) {
                $validResult = $this->msgCryptor->mpVerifySig($signature, $this->timestamp, $this->nonce);
            } else if (!empty($echostr)) {
                $validResult = $this->msgCryptor->qyVerifySig($this->msg_signature, $this->timestamp, $this->nonce, $echostr);
            }
            if (!$validResult) exit("ERR: 40001\n\n");
        }

        // 企业号解密echostr
        if($wechatType == 2 && !empty($echostr)) {
            $echostr = $this->msgCryptor->qyEchoStr($echostr);
        }

        // 如果是验证URL则直接输出
        if (!empty($echostr)) exit($echostr);

        if (empty($requestData)) exit('缺少数据');

        // 设置错误处理函数，将错误通过文本消息回复显示
        set_error_handler(array(&$this, 'errorHandler'));

        $this->savePostData($requestData);

    }

    /**
     * 解密并保存微信post的数据，保存至$this->request中
     *
     * @param string $requestData 微信post的数据
     * @return void
     */
    private function savePostData($requestData) 
    {
        
        if ($this->encrypted) {
            
            $xml = '';
            $errCode = $this->msgCryptor->decryptMsg($this->msg_signature, $this->timestamp, $this->nonce, $requestData, $xml);
            
            if ($errCode != 0) exit($errCode);

        } else {

            $xml = $requestData;

        }

        libxml_disable_entity_loader(true);
        $reqArray = (array) simplexml_load_string($xml, 'SimpleXMLElement', LIBXML_NOCDATA);
        if (empty($reqArray)) exit("xml解析失败");

        // 将数组键名转换为小写，提高健壮性，减少因大小写不同而出现的问题
        $this->request = array_change_key_case($reqArray, CASE_LOWER);

    }


    /**
      * 输出消息
      * @param string $msg 要输出的内容
      * @return void
      */
    private function echoMsg($msg)
    {
        if ($this->encrypted) {
            $sEncryptMsg = "";
            // 加密
            $errCode = $this->msgCryptor->EncryptMsg($msg, $this->timestamp, $this->nonce, $sEncryptMsg);
            if ($errCode != 0) {
                exit ($errCode);
            }
        } else {
            $sEncryptMsg = $msg;
        }
        echo $sEncryptMsg;
    }

    /**
      * 以文本消息回复
      * @param string $text 要回复的文本
      * @return void
      */
    public function echoText($text)
    {
        $resultStr = sprintf($this->templateText, 
            $this->request['fromusername'], 
            $this->request['tousername'], 
            time(), 
            trim($text)
        );
        $this->echoMsg($resultStr);
    }

    /**
      * 以图文消息回复
      * @param array $items 要回复的内容
      * @return void
      */
    public function echoNews($items)
    {
        $resultStr = sprintf($this->templateNewsBegin, 
            $this->request['fromusername'], 
            $this->request['tousername'], 
            time(), 
            count($items)
        );
        for ($i=0; $i<count($items); $i++) {
            $resultStr .=  sprintf($this->templateNewsContent, 
                isset($items[$i][0])?trim($items[$i][0]):'', 
                isset($items[$i][1])?trim($items[$i][1]):'', 
                isset($items[$i][2])?trim($items[$i][1]):'', 
                isset($items[$i][3])?trim($items[$i][1]):''
            );
        }
        $resultStr .=  $this->templateNewsEnd;
        $this->echoMsg($resultStr);
    }

    /**
      * 进行消息的识别及回复
      * @return void
      */
    public function run()
    {
        $msgType = strtolower($this->request['msgtype']);
        if (in_array($msgType, $this->msgTypeArray)) {
            eval('$this->respon_' . $msgType . '();');
        } else {
            $this->respon_unknown();
        }
    }

    // 文本消息
    public function respon_text() {}

    // 图片消息
    public function respon_image() {}

    // 语音消息
    public function respon_voice() {}

    // 视频消息
    public function respon_video() {}

    // 小视频消息
    public function respon_shortvideo() {}

    // 链接消息
    public function respon_link(){}

    // 位置消息
    public function respon_location() {}

    // 未知消息
    public function respon_unknown()
    {
        if (!$this->debug) {
            return;
        }

        $template = "PHP 报错啦！\r\n\r\n出现未知的消息类型%s";

        $content = sprintf($template, $this->request['event']);

        $this->echoText($content);
        exit;
    }

    // 事件消息
    public function respon_event()
    {
        $event = strtolower($this->request['event']);
        if (in_array($event, $this->eventTypeArray)) {
            eval('$this->respon_event_' . $event . '();');
        } else {
            $this->respon_event_unknown();
        }
    }

    // 订阅事件（公众号、企业号）
    public function respon_event_subscribe() {}

    // 取消订阅（公众号、企业号）
    public function respon_event_unsubscribe() {}

    // 点击自定义菜单（公众号、企业号）
    public function respon_event_click() {}

    // 进入自定义菜单网址（公众号、企业号）
    public function respon_event_view() {}

    // 自动上报地理位置（公众号、企业号）
    public function respon_event_location() {}

    // 扫描带参数二维码事件（公众号）
    public function respon_event_scan() {}

    // 扫码推事件的事件推送（企业号、公众号）
    public function respon_event_scancode_push() {}

    // 扫码推事件且弹出“消息接收中”（企业号、公众号）
    public function respon_event_scancode_waitmsg() {}

    // 弹出系统拍照发图（企业号、公众号）
    public function respon_event_pic_sysphoto() {}

    // 弹出拍照或者相册发图（企业号、公众号）
    public function respon_event_pic_photo_or_album() {}

    // 弹出微信相册发图器（企业号、公众号）
    public function respon_event_pic_weixin() {}

    // 弹出地理位置选择器（企业号、公众号）
    public function respon_event_location_select() {}

    // 成员进入应用（企业号）
    public function respon_event_enter_agent() {}

    // 异步任务完成（企业号）
    public function respon_event_batch_job_result() {}

    // 未知事件
    public function respon_event_unknown()
    {
        if (!$this->debug) {
            return;
        }

        $template = "PHP 报错啦！\r\n\r\n出现未知的事件类型%s";

        $content = sprintf($template, $this->request['event']);

        $this->echoText($content);
        exit;
    }

    /**
     * 自定义的错误处理函数，将 PHP 错误通过文本消息回复显示
     * @param  int $level   错误代码
     * @param  string $msg  错误内容
     * @param  string $file 产生错误的文件
     * @param  int $line    产生错误的行数
     * @return void
     */
    public function errorHandler($level, $msg, $file, $line) {
        if (!$this->debug) {
            return;
        }

        $error_type = array(
            // E_ERROR             => 'Error',
            E_WARNING           => 'Warning',
            // E_PARSE             => 'Parse Error',
            E_NOTICE            => 'Notice',
            // E_CORE_ERROR        => 'Core Error',
            // E_CORE_WARNING      => 'Core Warning',
            // E_COMPILE_ERROR     => 'Compile Error',
            // E_COMPILE_WARNING   => 'Compile Warning',
            E_USER_ERROR        => 'User Error',
            E_USER_WARNING      => 'User Warning',
            E_USER_NOTICE       => 'User Notice',
            E_STRICT            => 'Strict',
            E_RECOVERABLE_ERROR => 'Recoverable Error',
            E_DEPRECATED        => 'Deprecated',
            E_USER_DEPRECATED   => 'User Deprecated',
        );

        $template = "PHP 报错啦！\r\n\r\n%s: %s\r\nFile: %s\r\nLine: %s";

        $content = sprintf($template, $error_type[$level], $msg, $file, $line);

        $this->echoText($content);
        exit;
    }
    /**
     * 处理函数调用异常
     * @param  function $function_name   函数名
     * @param  string $args  参数列表
     * @return void
     */
    public function __call($function_name, $args) {
        if (!$this->debug) {
            return;
        }
        $template = "PHP 报错啦！\r\n\r\n函数%s不存在\r\n参数列表为：%s";

        $content = sprintf($template, $function_name, implode(',', $args));

        $this->echoText($content);
        exit;
    }
}

File: java/src/cn/kchen/jsp/src/Wechat.java
package cn.kchen.jsp.src;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.PrintWriter;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

import com.qq.weixin.mp.aes.AesException;

public class Wechat {

	public EventMessage mEventMessage;
	public String mKeyword;
	
    protected String mTemplateText        = "<xml><ToUserName><![CDATA[%s]]></ToUserName><FromUserName><![CDATA[%s]]></FromUserName><CreateTime>%s</CreateTime><MsgType><![CDATA[text]]></MsgType><Content><![CDATA[%s]]></Content></xml>";
    protected String mTemplateNewsBegin   = "<xml><ToUserName><![CDATA[%s]]></ToUserName><FromUserName><![CDATA[%s]]></FromUserName><CreateTime>%s</CreateTime><MsgType><![CDATA[news]]></MsgType><ArticleCount>%s</ArticleCount><Articles>";
    protected String mTemplateNewsContent = "<item><Title><![CDATA[%s]]></Title><Description><![CDATA[%s]]></Description><PicUrl><![CDATA[%s]]></PicUrl><Url><![CDATA[%s]]></Url></item>";
    protected String mTemplateNewsEnd     = "</Articles></xml>";
    protected String mTemplateCustomer    = "<xml><ToUserName><![CDATA[%s]]></ToUserName><FromUserName><![CDATA[%s]]></FromUserName><CreateTime>%s</CreateTime><MsgType><![CDATA[transfer_customer_service]]></MsgType></xml>";
    protected String mTemplateImage       = "<xml><ToUserName><![CDATA[%s]]></ToUserName><FromUserName><![CDATA[%s]]></FromUserName><CreateTime>%s</CreateTime><MsgType><![CDATA[image]]></MsgType><Image><MediaId><![CDATA[%s]]></MediaId></Image></xml>";
	
	private BufferedReader mBufferedReader;
	private PrintWriter mPrintWriter;
	private MsgCryptor mMsgCryptor;
    private String mTimestamp, mNonce, mMsgSignature;
    private boolean mEncrypted, mDebug;
	
	public Wechat(HttpServletRequest request, HttpServletResponse response, String token, String encodingAesKey, String appId, boolean debug) throws IOException, MyException{
		
        mDebug = debug;
        
		// 设置编码
		request.setCharacterEncoding("UTF-8");
        response.setCharacterEncoding("UTF-8");
        mBufferedReader = request.getReader();
        mPrintWriter = response.getWriter();

        // 检查参数
        mTimestamp = request.getParameter("timestamp");
        mNonce     = request.getParameter("nonce");
        if (!mDebug && (mTimestamp == null || mNonce == null)) {
        	throw new MyException("请通过微信访问");
        }

        // 判断消息是否加密
        mMsgSignature =  request.getParameter("msg_signature");
        mEncrypted = mMsgSignature != null;
        
        // 判断是公众号还是企业号（1.公众号；2.企业号。）
        String signature = request.getParameter("signature");
        int wechatType = (signature == null) ? 2 : 1;
        
        // 其他参数
        String echostr = request.getParameter("echostr");
        
        // 创建加密模块对象
        try {
            mMsgCryptor = new MsgCryptor(token, encodingAesKey, appId);
		} catch (Exception e) {
			e.printStackTrace();
        	throw new MyException("加密模块有误");
		}
        
        // 验证签名
        if(!mDebug) {
        	boolean validResult = true;
        	if (wechatType == 1) { // 公众号
        		validResult = mMsgCryptor.mpVerifySig(signature, mTimestamp, mNonce);
        	} else if (echostr != null) { // 企业号
            	validResult = mMsgCryptor.qyVerifySig(mMsgSignature, mTimestamp, mNonce, echostr);
        	}
        	if (!validResult) throw new MyException("签名验证失败");
        }

        // 解密并输出echostr
        if (echostr != null){
        	if (wechatType == 2) {
        		echostr = mMsgCryptor.qyEchoStr(mMsgSignature, mTimestamp, mNonce, echostr);
        	}
        	throw new MyException(echostr);
        }
        
        StringBuffer sb = new StringBuffer();
        String s;
        while((s = mBufferedReader.readLine()) != null) {
        	sb.append(s);
        }
        String xml = sb.toString();
        if (xml == null || xml.equals("")) throw new MyException("缺少数据");
        
        if (mEncrypted) {
        	try {
				xml = mMsgCryptor.decryptMsg(mMsgSignature, mTimestamp, mNonce, xml);
			} catch (AesException e) {
				e.printStackTrace();
				throw new MyException("消息体解密失败");
			}
        }
        
		mEventMessage = XMLConverUtil.convertToObject(EventMessage.class, xml);
	}
	
	private void echoMsg(String msg){
		if (mEncrypted) {
			try {
				msg = mMsgCryptor.encryptMsg(msg, mTimestamp, mNonce);
			} catch (AesException e) {
				e.printStackTrace();
			}
		}
        mPrintWriter.append(msg);
	}
	
	public void echoText(String text)
	{
		String resultStr = String.format(
				mTemplateText, 
				mEventMessage.getFromUserName(), 
				mEventMessage.getToUserName(), 
				System.currentTimeMillis()/1000,
				text);
		echoMsg(resultStr);
	}
	
	public void run()
	{	
		String MsgType = mEventMessage.getMsgType().toLowerCase();

		switch (MsgType) {
	        case "text":
	            mKeyword = mEventMessage.getContent();
	            respon_text();
	            break;
	        case "image":
	            mKeyword = mEventMessage.getPicUrl();
	            respon_image();
	            break;
	        case "voice":
	            mKeyword = mEventMessage.getRecognition();
	            respon_voice();
	            break;
	        case "video":
	            respon_video();
	            break;
	        case "shortvideo":
	            respon_shortvideo();
	            break;
	        case "link":
	            mKeyword = mEventMessage.getUrl();
	            respon_link();
	            break;
	        case "location":
	            mKeyword = mEventMessage.getLocation_X() + "." + mEventMessage.getLocation_Y();
	            respon_location();
	            break;
	        case "event":
	            respon_event();
	            break;
	        default:
	            respon_unknown();
	            break;
		}
	}

    // 文本消息
	public void respon_text() {}

    // 图片消息
	public void respon_image() {}

    // 语音消息
	public void respon_voice() {}

    // 视频消息
	public void respon_video() {}

    // 小视频消息
	public void respon_shortvideo() {}

    // 链接消息
	public void respon_link() {}

    // 位置消息
	public void respon_location() {}

    // 未知消息
	public void respon_unknown() {
        if (!mDebug) {
            return;
        }
        String template = "Servlet 报错啦！\r\n\r\n出现未知的消息类型%s";
        String content = String.format(template, mEventMessage.getEvent());
        echoText(content);
	}

    // 事件消息
	public void respon_event() {
		String event = mEventMessage.getEvent().toLowerCase();
		switch (event) {
	        case "subscribe":
	            respon_event_subscribe();
	            break;
	        case "unsubscribe":
	            respon_event_unsubscribe();
	            break;
	        case "click":
	            mKeyword = mEventMessage.getEventKey();
	            respon_event_click();
	            break;
	        case "view":
	            mKeyword = mEventMessage.getEventKey();
	            respon_event_view();
	            break;
	        case "location":
	            respon_event_location();
	            break;
	        case "scan":
	            respon_event_scan();
	            break;
	        case "scancode_push":
	            respon_event_scancode_push();
	            break;
	        case "scancode_waitmsg":
	            respon_event_scancode_waitmsg();
	            break;
	        case "pic_sysphoto":
	            respon_event_pic_sysphoto();
	            break;
	        case "pic_photo_or_album":
	            respon_event_pic_photo_or_album();
	            break;
	        case "pic_weixin":
	            respon_event_pic_weixin();
	            break;
	        case "location_select":
	            respon_event_location_select();
	            break;
	        case "enter_agent":
	            respon_event_enter_agent();
	            break;
	        case "batch_job_result":
	            respon_event_batch_job_result();
	        case "templatesendjobfinish":
	            respon_event_templatesendjobfinish();
	            break;
	        
	        default:
	            respon_event_unknown();
	            break;
		}
	}

    // 订阅事件（公众号、企业号）
    public void respon_event_subscribe() {}

    // 取消订阅（公众号、企业号）
    public void respon_event_unsubscribe() {}

    // 点击自定义菜单（公众号、企业号）
    public void respon_event_click() {}

    // 进入自定义菜单网址（公众号、企业号）
    public void respon_event_view() {}

    // 自动上报地理位置（公众号、企业号）
    public void respon_event_location() {}

    // 扫描带参数二维码事件（公众号）
    public void respon_event_scan() {}

    // 扫码推事件的事件推送（企业号、公众号）
    public void respon_event_scancode_push() {}

    // 扫码推事件且弹出“消息接收中”（企业号、公众号）
    public void respon_event_scancode_waitmsg() {}

    // 弹出系统拍照发图（企业号、公众号）
    public void respon_event_pic_sysphoto() {}

    // 弹出拍照或者相册发图（企业号、公众号）
    public void respon_event_pic_photo_or_album() {}

    // 弹出微信相册发图器（企业号、公众号）
    public void respon_event_pic_weixin() {}

    // 弹出地理位置选择器（企业号、公众号）
    public void respon_event_location_select() {}

    // 成员进入应用（企业号）
    public void respon_event_enter_agent() {}

    // 异步任务完成（企业号）
    public void respon_event_batch_job_result() {}

    // 模板消息的事件推送（公众号）
    public void respon_event_templatesendjobfinish() {}

    // 未知事件
    public void respon_event_unknown()
    {
        if (!mDebug) {
            return;
        }
        String template = "Servlet 报错啦！\r\n\r\n出现未知的事件类型%s";
        String content = String.format(template, mEventMessage.getEvent());
        echoText(content);
    }
}
